### PR TITLE
nvidia-x11: rewrite build

### PR DESCRIPTION
**Pull request by codyopel**
nvidia-x11: rewrite build
- Support for version 361+
  - NVIDIA has added experimental support for EGL, bringing wayland
    support closer
  - Support for NVIDIA's new GLVND libraries
- Fix i686 support (for whatever thats worth)
  - Cuda UVM is only supported on x86_64 for 346+
- Fix kernel modules for some platforms & versions
- Use the long-lived driver as the default
  - Previously it was unclear what the practice is for the default.
- Add long-lived, short-lived, beta, vulkan drivers
  - Available as "nvidia/nvidiaLongLived", "nvidiaShortLived",
    "nvidiaBeta", & "nvidiaVulkan" in services.xserver.videoDrivers
- Remove support for 173.xx
  - This driver branch requires a version of xorg-server that is no longer
    supported in nixpkgs.
- Fix patchelf allowing --shrink-rpath to be used
- Install everything explicitly to catch build failures due to changes
- Cleanup shared object version symlinking, _.so_
- Cuda fixes
  - Add missing cuda executables, libraries and manpages
- GTK+3 support and used by default (GTK+3 is supported for 346+)
  - nvidia-settings now works correctly on high dpi displays
- Add additional separation of kernelspace and userspace to fix packaging of
  packages building against the nvidia-drivers. This is currently incomplete,
  but the ground work for doing so is complete.
- Add a consistent source unpack method
- Add nvidia default application profiles
- Add better documentation to builder
- Add X.Org driver config
- Add NVIDIA ICD config
- Cleanup buildinputs (add missing deps & remove unused)

cc @vcunat
After taking a quick look at the current build I didn't want to decipher what is broken and what isn't. This is basically copied verbatim from my build in https://github.com/triton/triton/tree/master/pkgs/all-pkgs/nvidia-drivers, so it contain fixes for all branches and platforms.  I don't currently have a system to test this on, but I believe I fixed all attribute names so that it should just work within nixpkgs.

### CODE REVIEW

**Comment by codyopel** [comment]
Related discussion:
https://github.com/NixOS/nixpkgs/pull/12719
Resolves issue:
https://github.com/NixOS/nixpkgs/issues/13942

**Comment by codyopel** [comment]
@vcunat
Also, all branches build except the beta branch because NVIDIA renamed `libGL.so.1` -> `libGL.so.1.0.0` in a point release and the generic builder is based on using major versions for such changes.

**Comment by vcunat** [comment]
We should patch out the "triton" strings, I think... but I only skimmed all this for now. I'll try to do a better review later today or tomorrow, unless someone beats me.

File: nixos/modules/hardware/video/nvidia.nix [hunk 1]
```diff
@@ -2,7 +2,14 @@
...
 
-with lib;
+with {
+  inherit (lib)
+    elem
+    mkIf
+    optionals
+    optionalString
+    singleton;
+};
```
Comment (edolstra) [inline nixos/modules/hardware/video/nvidia.nix]: Please do don't this, it's unnecessarily verbose.

**Comment by edolstra** [comment]
I'm concerned about the maintainability of a nearly 600-line build script, which is nearly quadruple the size of the previous builder.sh. Where does that increase come from?

GTK+3 is still not the default in Nixpkgs, so I don't think programs should use it if they can also use GTK+2, in order not to end up with multiple versions of GTK+3 in the system closure.

`pkgs/os-specific/linux/nvidia-x11` should use `callPackages` rather than relying on a `channel` argument.

**Comment by codyopel** [comment]
@edolstra 
I only opened the PR at the request of @vcunat . If it isn't what you want you can manually add the fixes from my changes. I detailed the necessary changes in a linked PR that will resolve the errors, because as it stands there are packaging errors that will cause the driver to segfault. These changes do not effect me, so they are what they are, take it or leave it.  Also the builder is longer because it is a generic builder for all versions, that also includes better tests for catching errors.

lol, in 2016 you still prefer gtk2 as the default, have fun with that

**Comment by vcunat** [comment]
I see this won't be as easy as I hoped, so I reverted that update for now to avoid crashes until we find a working and "acceptable" solution.

**Comment by codyopel** [comment]
changes in 364.12 beta: https://github.com/triton/triton/commit/acc4ffa9ed267d1043e2f30426c55eb295799e8a

**Comment by domenkozar** [comment]
@vcunat does nvidia work now on 16.03 without this PR?

**Comment by vcunat** [comment]
Yes, as I posted, for now I had reverted the branches to 358.16, as that version worked reliably and it really isn't that old.

**Comment by dezgeg** [comment]
Sounds like the issue got solved some other way, closing.

**Pull request closed without merging**